File: backend/shared/core/src/main/java/io/mateu/core/application/mcp/McpService.java
package io.mateu.core.application.mcp;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import io.mateu.core.application.MateuService;
import io.mateu.dtos.RunActionRqDto;
import io.mateu.dtos.UIIncrementDto;
import io.mateu.uidl.interfaces.HttpRequest;
import java.util.Map;

/**
 * The native (in-process) host of the agent-operability plane: turns Model Context Protocol tool
 * calls into {@link MateuService} sync requests and projects the resulting wire with the SAME
 * {@link McpProjection} the sidecar uses. RBAC is enforced by {@code MateuService} over the passed
 * {@link HttpRequest} (the JWT), so an action a caller may not run never reaches the projection.
 *
 * <p>Plain class (constructed by the adapter controller and by tests), so it adds nothing to the
 * bean graph. See {@code design/riu-agent-operability-plan.md} (P3).
 */
public class McpService {

  /**
   * Builds the concrete {@link HttpRequest} for a request from the (framework-specific) adapter —
   * the mvc controller wraps the servlet request (carrying the JWT for RBAC), a test wraps a fake.
   * Kept as a factory so core stays framework-neutral and the rq stored on the request matches the
   * rq passed to {@code runAction}.
   */
  public interface HttpRequestFactory {
    HttpRequest create(RunActionRqDto rq);
  }

  private final MateuService service;
  private final ObjectMapper objectMapper;

  public McpService(MateuService service, ObjectMapper objectMapper) {
    this.service = service;
    this.objectMapper = objectMapper;
  }

  /** Load a screen (route) → flat projection. The load is dispatched as actionId "". */
  public ObjectNode describeScreen(String route, String baseUrl, HttpRequestFactory httpRequests) {
    return project(sync(rq(route, "", null, null), baseUrl, httpRequests));
  }

  /** Run an action on a screen → resulting projection. */
  public ObjectNode runAction(
      String route,
      String actionId,
      Map<String, Object> componentState,
      String baseUrl,
      HttpRequestFactory httpRequests) {
    return project(sync(rq(route, actionId, componentState, null), baseUrl, httpRequests));
  }

  /** Run the standard {@code search} action of a listing. */
  public ObjectNode search(
      String route,
      String searchText,
      Map<String, Object> filters,
      String baseUrl,
      HttpRequestFactory httpRequests) {
    Map<String, Object> params = searchText == null ? Map.of() : Map.of("searchText", searchText);
    return project(sync(rq(route, "search", filters, params), baseUrl, httpRequests));
  }

  /** Navigable routes from the app menu on the root response → [{route, caption}]. */
  public ArrayNode listRoutes(String baseUrl, HttpRequestFactory httpRequests) {
    JsonNode increment = sync(rq("", "", null, null), baseUrl, httpRequests);
    java.util.LinkedHashMap<String, String> routes = new java.util.LinkedHashMap<>();
    collectRoutes(increment.path("fragments"), routes);
    routes.putIfAbsent("", "Home");
    ArrayNode out = objectMapper.createArrayNode();
    routes.forEach(
        (route, caption) -> {
          ObjectNode o = out.addObject();
          o.put("route", route);
          o.put("caption", caption);
        });
    return out;
  }

  private void collectRoutes(JsonNode node, Map<String, String> out) {
    if (node == null) return;
    if (node.isObject()) {
      JsonNode md = node.path("metadata");
      if (md.isObject() && "App".equals(md.path("type").asText(null)))
        walkMenu(md.path("menu"), out);
      if (node.hasNonNull("route") && !node.get("route").asText().isEmpty()) {
        String r = McpProjection.normalizeRoute(node.get("route").asText());
        out.putIfAbsent(
            r,
            node.hasNonNull("caption") ? node.get("caption").asText() : (r.isEmpty() ? "Home" : r));
      }
      for (JsonNode child : node) collectRoutes(child, out);
    } else if (node.isArray()) {
      for (JsonNode child : node) collectRoutes(child, out);
    }
  }

  private void walkMenu(JsonNode menu, Map<String, String> out) {
    if (!menu.isArray()) return;
    for (JsonNode item : menu) {
      JsonNode routeNode = item.hasNonNull("route") ? item.get("route") : item.path("path");
      if (routeNode != null && routeNode.isTextual() && !routeNode.asText().isEmpty()) {
        String r = McpProjection.normalizeRoute(routeNode.asText());
        out.putIfAbsent(
            r,
            item.hasNonNull("caption") ? item.get("caption").asText() : (r.isEmpty() ? "Home" : r));
      }
      walkMenu(item.path("submenus"), out);
      walkMenu(item.path("menu"), out);
    }
  }

  private RunActionRqDto rq(
      String route,
      String actionId,
      Map<String, Object> componentState,
      Map<String, Object> parameters) {
    // MateuService resolves a mount route in leading-slash form ("/mcp-demo"); the root stays "".
    String normalized = McpProjection.normalizeRoute(route);
    String rqRoute = normalized.isEmpty() ? "" : "/" + normalized;
    return RunActionRqDto.builder()
        .route(rqRoute)
        .actionId(actionId == null ? "" : actionId)
        .componentState(componentState == null ? Map.of() : componentState)
        .appState(Map.of())
        .parameters(parameters == null ? Map.of() : parameters)
        .build();
  }

  private JsonNode sync(RunActionRqDto rq, String baseUrl, HttpRequestFactory httpRequests) {
    try {
      UIIncrementDto increment =
          service.runAction(baseUrl, rq, baseUrl, httpRequests.create(rq)).next().block();
      return objectMapper.valueToTree(increment);
    } catch (Throwable t) {
      throw new RuntimeException(
          "MCP sync failed for route '" + rq.route() + "': " + t.getMessage(), t);
    }
  }

  private ObjectNode project(JsonNode increment) {
    return McpProjection.project(increment, objectMapper);
  }
}

File: backend/dotnet/src/Mateu.Core/Mcp/McpService.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using Mateu.Dtos;

namespace Mateu.Core.Mcp;

/// <summary>
/// The native (in-process) host of the agent-operability plane: turns MCP tool calls into
/// <see cref="SyncHandler"/> requests and projects the resulting wire with the same
/// <see cref="McpProjection"/> the sidecar uses. RBAC is enforced by the handler (an [EyesOnly]
/// member a caller may not see never reaches the projection). C# mirror of the Java McpService.
/// </summary>
public sealed class McpService
{
    readonly SyncHandler _handler;
    readonly JsonSerializerOptions _json;
    readonly string? _requestBaseUrl;

    public McpService(SyncHandler handler, JsonSerializerOptions json, string? requestBaseUrl = null)
    {
        _handler = handler;
        _json = json;
        _requestBaseUrl = requestBaseUrl;
    }

    /// <summary>Load a screen (route) → flat projection. The load is dispatched as actionId "".</summary>
    public JsonObject DescribeScreen(string route) =>
        McpProjection.Project(Sync(Rq(route, "", null, null)));

    /// <summary>Run an action on a screen → resulting projection.</summary>
    public JsonObject RunAction(string route, string actionId, JsonObject? componentState) =>
        McpProjection.Project(Sync(Rq(route, actionId, componentState, null)));

    /// <summary>Run the standard <c>search</c> action of a listing.</summary>
    public JsonObject Search(string route, string? searchText, JsonObject? filters)
    {
        JsonObject? parameters = null;
        if (!string.IsNullOrEmpty(searchText)) parameters = new JsonObject { ["searchText"] = searchText };
        return McpProjection.Project(Sync(Rq(route, "search", filters, parameters)));
    }

    /// <summary>Navigable routes from the app menu on the root response → [{route, caption}].</summary>
    public JsonArray ListRoutes()
    {
        var increment = Sync(Rq("", "", null, null));
        var routes = new List<KeyValuePair<string, string>>();
        var seen = new HashSet<string>();
        void Add(string r, string caption) { if (seen.Add(r)) routes.Add(new(r, caption)); }
        Collect(increment?["fragments"] as JsonArray, Add);
        if (!seen.Contains("")) routes.Add(new("", "Home"));
        var arr = new JsonArray();
        foreach (var kv in routes) arr.Add(new JsonObject { ["route"] = kv.Key, ["caption"] = kv.Value });
        return arr;
    }

    static void Collect(JsonNode? node, Action<string, string> add)
    {
        if (node is JsonObject obj)
        {
            if (obj["metadata"] is JsonObject md && AsString(md["type"]) == "App")
                WalkMenu(md["menu"], add);
            if (obj["route"] is JsonValue rv && rv.TryGetValue<string>(out var route) && route.Length > 0)
            {
                var r = McpProjection.NormalizeRoute(route);
                add(r, AsString(obj["caption"]) ?? AsString(obj["label"]) ?? (r.Length == 0 ? "Home" : r));
            }
            foreach (var kv in obj) Collect(kv.Value, add);
        }
        else if (node is JsonArray arr)
        {
            foreach (var item in arr) Collect(item, add);
        }
    }

    static void WalkMenu(JsonNode? menu, Action<string, string> add)
    {
        if (menu is not JsonArray items) return;
        foreach (var itemEl in items)
        {
            if (itemEl is not JsonObject item) continue;
            var raw = AsString(item["route"]) ?? AsString(item["path"]);
            if (!string.IsNullOrEmpty(raw))
            {
                var r = McpProjection.NormalizeRoute(raw);
                add(r, AsString(item["caption"]) ?? AsString(item["label"]) ?? (r.Length == 0 ? "Home" : r));
            }
            WalkMenu(item["submenus"], add);
            WalkMenu(item["menu"], add);
        }
    }

    static string? AsString(JsonNode? n) => n is JsonValue v && v.TryGetValue<string>(out var s) ? s : null;

    RunActionRqDto Rq(string route, string actionId, JsonObject? componentState, JsonObject? parameters)
    {
        // SyncHandler resolves a mount route in leading-slash form ("/mcp-demo"); root stays "".
        var normalized = McpProjection.NormalizeRoute(route);
        return new RunActionRqDto
        {
            Route = normalized.Length == 0 ? "" : "/" + normalized,
            ActionId = actionId ?? "",
            ComponentState = ToDict(componentState),
            Parameters = ToDict(parameters),
        };
    }

    // Convert a JsonObject to the Dictionary<string,object?> shape SyncHandler expects — values as
    // JsonElement, exactly as the HTTP sync endpoint deserializes componentState.
    Dictionary<string, object?> ToDict(JsonObject? obj)
    {
        if (obj is null) return new();
        return JsonSerializer.Deserialize<Dictionary<string, object?>>(obj.ToJsonString(), _json) ?? new();
    }

    JsonNode? Sync(RunActionRqDto rq)
    {
        var increment = _handler.Handle(rq, _requestBaseUrl);
        return JsonSerializer.SerializeToNode(increment, _json);
    }
}
